Sources/SwiftMemcache/MemcachedConnection.swift

```diff
@@ -0,0 +1,170 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-memcache-gsoc open source project
+//
+// Copyright (c) 2023 Apple Inc. and the swift-memcache-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-memcache-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+@_spi(AsyncChannel)
+
+import NIOCore
+import NIOPosix
+
+/// An actor to create a connection to a Memcache server.
+///
+/// This actor can be used to send commands to the server.
+public actor MemcachedConnection {
+    private typealias StreamElement = (MemcachedRequest, CheckedContinuation<MemcachedResponse, Error>)
+    /// The underlying channel to communicate with the server.
+    private let channel: NIOAsyncChannel<MemcachedResponse, MemcachedRequest>
+    /// The channel's event loop group.
+    private let eventLoopGroup: EventLoopGroup
+
+    /// Enum representing the current state of the MemcachedConnection.
+    ///
+    /// The ConnectionState is either running or finished, depending on whether the connection
+    /// to the server is active or has been closed. When running, it contains the properties
+    /// for the buffer allocator, request stream, and the stream's continuation.
+    private enum ConnectionState {
+        case running(
+            /// The allocator used to create new buffers.
+            bufferAllocator: ByteBufferAllocator,
+            /// The stream of requests to be sent to the server.
+            requestStream: AsyncStream<StreamElement>,
+            /// The continuation for the request stream.
+            requestContinuation: AsyncStream<StreamElement>.Continuation
+        )
+        case finished
+    }
+
+    /// Enum representing the possible errors that can be encountered in `MemcachedConnection`.
+    enum MemcachedConnectionError: Error {
+        /// Indicates that the connection to the server has finished.
+        case connectionFinished
+        /// Indicates that the request has been dropped because the buffer was full.
+        case requestDropped
+        /// Indicates that the request was not enqueued because the stream was in a terminal state.
+        case requestStreamTerminated
```

Comment:
All of those are a single error basically. Can we just call that case `connectionShutdown` for now